numpy/core/src/multiarray/conversion_utils.c

```diff
@@ -162,6 +164,82 @@ PyArray_OptionalIntpConverter(PyObject *obj, PyArray_Dims *seq)
     return PyArray_IntpConverter(obj, seq);
 }
 
+NPY_NO_EXPORT int
+PyArray_CopyConverter(PyObject *obj, PyArray_CopyMode *copymode) {
+    if (obj == Py_None) {
+        PyErr_SetString(PyExc_ValueError,
+                        "NoneType copy mode not allowed. Please choose one of " 
+                         "np.CopyMode.ALWAYS, np.CopyMode.IF_NEEDED, np.CopyMode.NEVER.");
+        return NPY_FAIL;
+    }
+
+    int int_copymode = -1;
+    PyObject* numpy_CopyMode = NULL;
+    PyObject* numpy_bool_ = NULL;
+    npy_cache_import("numpy._globals", "CopyMode", &numpy_CopyMode);
+    npy_cache_import("numpy", "bool_", &numpy_bool_);
+
+    if (numpy_CopyMode != NULL && PyObject_IsInstance(obj, numpy_CopyMode)) {
+        PyObject* mode_value = PyObject_GetAttrString(obj, "value");
+        if (mode_value == NULL) {
+            return NPY_FAIL;
+        }
+        if (!PyArray_PythonPyIntFromInt(mode_value, &int_copymode)) {
+            return NPY_FAIL;
+        }
+    } else if(numpy_bool_ != NULL && PyObject_IsInstance(obj, numpy_bool_)) {
+        /*
+        In Python 3.8.2, numpy.bool_ objects are not converted
+        to integers using PyLong_AsLong. Hence, the following
+        if condition checks whether the input obj is a numpy.bool_ object
+        and extracts the values by perfroming comparisions between
+        obj and numpy.True_/numpy.False_.
+        */
+        PyObject *numpy_True_ = NULL, *numpy_False_ = NULL;
+        npy_cache_import("numpy", "True_", &numpy_True_);
+        npy_cache_import("numpy", "False_", &numpy_False_);
+        if( numpy_True_ != NULL && 
+            PyObject_RichCompareBool(obj, numpy_True_, Py_EQ) ) {
+            int_copymode = 1;
+        } else if( numpy_False_ != NULL && 
+                   PyObject_RichCompareBool(obj, numpy_False_, Py_EQ) ) {
+            int_copymode = 0;
+        }
+    } else if( PyBool_Check(obj) ) {
+        /*
+        In Python 3.8.2, Boolean objects are not converted
+        to integers using PyLong_AsLong. Hence, the following
+        if condition checks whether the input obj is a Boolean object
+        and extracts the values by perfroming comparisions between
+        obj and Py_True/Py_False.
+        */
+        if( PyObject_RichCompareBool(obj, Py_True, Py_EQ) ) {
+            int_copymode = 1;
+        } else if( PyObject_RichCompareBool(obj, Py_False, Py_EQ) ) {
+            int_copymode = 0;
+        }
+    }
+    else {
+        // If obj is not an instance of numpy.CopyMode then follow
+        // the conventional assumption that it must be value that
+        // can be converted to an integer.
+        PyArray_PythonPyIntFromInt(obj, &int_copymode);
```

Comment:
This seems a little messier than I would guess we need.

Can we just do the equivalent of `bool(copy_mode)` and then convert into the integer?

I don't think we need to preserve the legacy path which passes on integers unchanged. That was just a convenient implementation choice before, back when the only valid values of `copy_mode` were `True` and `False`.

Reply:
Actually, `PyArray_PythonPyIntFromInt` does, `int(copy_mode)`. However, this doesn't work in Python 3.8.2, so I resorted to the above method. I will try if something like `bool(copy_mode)` works.